File: csharp/sdk/Sdk/Event/TaskEventCollector.cs
using System.Net;
using Newtonsoft.Json.Linq;
using static System.Console;

namespace Acc.Service.Sdk.Event;

using Client;
using Types;

public interface ITaskEventListener
{
    void OnNewPaymentEvent(
        string type,
        int code,
        string message,
        long sequence,
        PaymentTaskItem paymentTaskItem
    );

    void OnNewShopEvent(
        string type,
        int code,
        string message,
        long sequence,
        ShopTaskItem shopTaskItem
    );
}

public class TaskEventCollector(PaymentClient client, ITaskEventListener listener) : Scheduler
{
    private long _sequence = 0;

    protected override async Task OnStart() 
    {
        WriteLine("TaskEventCollector:OnStart");
        try {
            this._sequence = await client.GetLatestTaskSequence();
            WriteLine($"Received sequence = {this._sequence}");
        } catch (Exception) {
            //
        }
    }

    protected override async Task OnWork() 
    {
        try
        {
            var tasks = await client.GetTasks(this._sequence);
            
            foreach (var t in tasks)
            {
                var dataType = t["type"]!.ToString();
                var code = Convert.ToInt32(t["code"]!.ToString());
                var message = t["message"]!.ToString();
                var sequence = Convert.ToInt64(t["sequence"]!.ToString());
                if (sequence > this._sequence) this._sequence = sequence;
                var data = t["data"]!.ToObject<JObject>();
                if (data == null) continue;
                if (dataType.Equals("pay_new") || dataType.Equals("pay_cancel")) {
                    var payment = PaymentTaskItem.FromJObject(data);
                    listener.OnNewPaymentEvent(
                        dataType,
                        code,
                        message,
                        sequence,
                        payment
                    );
                } else {
                    var shop = ShopTaskItem.FromJObject(data);
                    listener.OnNewShopEvent(
                        dataType,
                        code,
                        message,
                        sequence,
                        shop
                    );
                }
            }
            Thread.Sleep(2000);
        } catch (Exception) {
            //
        }
    }

    protected override async Task OnStop()
    {
        await Task.Delay(0);
        WriteLine("TaskEventCollector:OnStop");
    }
}

File: java/sdk/src/main/java/org/acc/service/sdk/event/TaskEventCollector.java
package org.acc.service.sdk.event;

import org.acc.service.sdk.client.PaymentClient;
import org.acc.service.sdk.data.payment.PaymentTaskItem;
import org.acc.service.sdk.data.payment.ShopTaskItem;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Objects;

public class TaskEventCollector extends Scheduler {

    private PaymentClient client = null;
    private ITaskEventListener listener = null;
    private long sequence = 0;

    public TaskEventCollector(PaymentClient client, ITaskEventListener listener) {
        this.client = client;
        this.listener = listener;
    }

    public void onStart() {
        System.out.println("TaskEventCollector:onStart");
        try {
            this.sequence = this.client.getLatestTaskSequence();
            System.out.printf("Received sequence = %d\n", this.sequence);
        } catch (Exception ignored) {

        }
    }

    public void onWork() {
        try {
            JSONArray tasks = this.client.getTasks(this.sequence);
            for (int idx = 0; idx < tasks.length(); idx++) {
                JSONObject task = tasks.getJSONObject(idx);
                String dataType = task.getString("type");
                int code = task.getInt("code");
                String message = task.getString("message");
                long sequence = task.getLong("sequence");
                if (sequence > this.sequence) this.sequence = sequence;
                System.out.printf("Received sequence = %d\n", this.sequence);

                JSONObject data = task.getJSONObject("data");
                if (Objects.equals(dataType, "pay_new") || Objects.equals(dataType, "pay_cancel")) {
                    PaymentTaskItem payment = PaymentTaskItem.fromJSONObject(data);
                    this.listener.onNewPaymentEvent(
                            dataType,
                            code,
                            message,
                            sequence,
                            payment
                    );
                } else {
                    ShopTaskItem shop = ShopTaskItem.fromJSONObject(data);
                    this.listener.onNewShopEvent(
                            dataType,
                            code,
                            message,
                            sequence,
                            shop
                    );
                }
            }
            Thread.sleep(2000);
        } catch (Exception ignored) {

        }
    }

    public void onStop() {
        System.out.println("TaskEventCollector:onStop");
    }
}
